### ISSUE TITLE
automating labeling of broken packages

### ISSUE THREAD

**Issue by daler** [bioconda/bioconda-utils#232]
(continuing from https://github.com/bioconda/bioconda-recipes/issues/6637#issuecomment-348510372, cc @johanneskoester @mbargull)

To summarize https://github.com/bioconda/bioconda-recipes/issues/6637:

Currently, when we find a package to be broken, we bump the build number. Our reasoning for this was that conda would always try to install the latest build number and ignore earlier build numbers. However, this is not the case, and some situations conda installs earlier build numbers. Instead, we need to mark a package as broken so that it will not be installed. We can do this by adding the `broken` [label](https://docs.anaconda.com/anaconda-repository/2.32/user/using#using-labels-in-the-development-cycle) to the package instead of the default `main` label.

It would be nice to do this via bioconda-utils on travis, so that contributors could mark their packages themselves without requiring core members to manually mark it. Furthermore, it would also be useful to have a record of this in the recipe somehow in git.

I suspect this will be another key in `extra:`. What about:

```yaml
name: examplepkg
version: 1.0
build:
  number: 4
...
extra:
  add_labels:
    broken:
      - 
        version: 1.0
        build_number: 2
      - 
        version: 1.0
        build_number: 3
```

When merged into master, bioconda-utils will mark `examplepkg-1.0_2` and `examplepkg-1.0_3` as broken by adding the `broken` label if that is not already the case. If, in the future, we determine that build 3 should not be considered broken, they can be removed from the list and by doing so it will be re-labeled as the default `main` label:

```yaml
name: examplepkg
version: 1.0
build:
  number: 4
...
extra:
  add_labels:
    broken:
      - 
        version: 1.0
        build_number: 2
```

Furthermore, we can add additional labels using this mechanism. For example, our channel also currently has a `deprecated` label, which could be added the same way:
```yaml
name: examplepkg
version: 1.0
build:
  number: 4
...
extra:
  add_labels:
    broken:
      - 
        version: 1.0
        build_number: 2
    deprecated:
      - 
        version: 1.0
        build_number: 3
```

**Comment by johanneskoester** [comment]
Thanks for working on this! I don't think this should be part of the recipe though. The reason is that we will have to mark stuff that is not anymore in the master branch. Eg the recipe is gone to conda forge. Or, in cases where multiple versions of a recipe are in subfolders, it is unclear which to choose for the label annotation. 

Instead, I would say a central tsv file with one column for the PKG including version and build string in the format expected by conda and the second column with the label would be preferable.

**Comment by daler** [comment]
Good points -- central TSV it is. I'd also add a third column for "reason".

**Comment by johanneskoester** [comment]
In fact, it should also contain the OS. Might be that a package is only broken on one system.

**Comment by mbargull** [comment]
I believe one central `.tsv` will become too unwieldy quickly.
One issue is that it will grow rather fast, e.g., see the possible entries the recent labelling of `busco` builds would create:
```tsv
linux-64/busco-3.0.2-py35_3.tar.bz2	broken	erroneous post-link.sh
linux-64/busco-3.0.2-py36_3.tar.bz2	broken	erroneous post-link.sh
linux-64/busco-3.0.2-py35_2.tar.bz2	broken	missing dependencies; erroneous post-link.sh
linux-64/busco-3.0.2-py36_2.tar.bz2	broken	missing dependencies; erroneous post-link.sh
linux-64/busco-3.0.2-py35_1.tar.bz2	broken	missing dependencies
linux-64/busco-3.0.2-py36_1.tar.bz2	broken	missing dependencies
linux-64/busco-3.0.2-py35_0.tar.bz2	broken	missing dependencies
linux-64/busco-3.0.2-py36_0.tar.bz2	broken	missing dependencies 
```
Add another 33 % of lines if it were Python 2 compatible and double them if there were builds for `osx-64`, too.
The second issue would be merge conflicts, of course.

Beware that such a file, contrary to files like `build-fail-blacklist`, may very well be in written to comparatively frequently. Though, *hopefully* not as frequently such that merge conflicts would happen that often.
[Granted, I may have a near-irrational aversion to large (and possibly un-/semi-structured) text files as (central) information storage, but still...]

**Comment by daler** [comment]
@mbargull I'm not too worried about size/complexity as TSV tooling is pretty solid. That said, maybe we can support wildcards/globs/regex in the package names to handle wide swaths of broken packages. So your example could be:

```
linux-64/busco-3.0.2-py3{5,6}_3.tar.bz2	       broken   erroneous post-link.sh
linux-64/busco-3.0.2-py3{5,6}_2.tar.bz2	       broken   missing dependencies; erroneous post-link.sh
linux-64/busco-3.0.2-py3{5,6}_{0,1}.tar.bz2    broken	missing dependencies
```

**Comment by mbargull** [comment]
I was thinking about cases when someone were to be confronted with it in a purely text-based manner like on merges.
But I like your idea of using globs or regular expressions to reduce the redundancy! (preferably globs -- by KISS principle)

**Comment by johanneskoester** [comment]
@daler, do you want to implement this? Otherwise, I think @mbargull could do it as well.

**Comment by daler** [comment]
Sorry, I'm swamped and won't have time to work on this for a while -- @mbargull's been doing great things, might as well add this to his list!

**Comment by mbargull** [comment]
Most likely I won't get to do this the coming two weeks. But maybe I can look into this over the holidays (to avoid some Xmas festivities -- call me Grinch ;P). I think I'll talk with Johannes about this tomorrow.

**Comment by mbargull** [comment]
Hmm, looks like `anaconda-client` only has `upload`/`copy` commands which support labels but none that lets us add/remove/set/move to some label.
Hence we might be blocked by https://github.com/Anaconda-Platform/anaconda-client/issues/351 on this.

**Comment by epruesse** [comment]
What's the status here? Do we close this?

I suppose with progress on the conda side and Bioconda "getting the CDN treatment" now, the best option to deal with broken packages it to actually remove them from the `repodata.json` (which I understand we now can).

**Comment by daler** [comment]
Agreed; the TSV described above would, in the end, be effectively making `repodata.json` edits of some form so it would be best to do that more directly.
